File: internal/application/commands/verify_lore.go
package commands

import (
	"context"
	"fmt"
	"strings"

	"github.com/memlore/memlore/internal/application/ports"
	"github.com/memlore/memlore/internal/domain"
)

// VerifyLoreCommand marks a lore entry verified.
type VerifyLoreCommand struct {
	EntryID string
	ActorID string
}

// VerifyLoreHandler handles lore verification.
type VerifyLoreHandler struct {
	begin ports.UnitOfWorkFactory
	clock ports.Clock
}

func NewVerifyLoreHandler(begin ports.UnitOfWorkFactory, clock ports.Clock) *VerifyLoreHandler {
	return &VerifyLoreHandler{begin: begin, clock: clock}
}

func (h *VerifyLoreHandler) Handle(ctx context.Context, cmd VerifyLoreCommand) (domain.LoreEntry, error) {
	actor := strings.TrimSpace(cmd.ActorID)
	if actor == "" {
		return domain.LoreEntry{}, &domain.ValidationError{Message: "actor must be non-empty"}
	}

	uow, err := h.begin(ctx)
	if err != nil {
		return domain.LoreEntry{}, err
	}
	defer uow.Rollback(ctx)

	entry, err := uow.LoreEntries().Get(ctx, cmd.EntryID)
	if err != nil {
		var nf *domain.NotFoundError
		if !isNotFound(err, &nf) {
			return domain.LoreEntry{}, err
		}
		return domain.LoreEntry{}, &domain.NotFoundError{
			Message: fmt.Sprintf("lore entry %s not found", cmd.EntryID),
		}
	}

	updated, audit, err := domain.ApplyVerification(entry, actor, h.clock.Now())
	if err != nil {
		return domain.LoreEntry{}, err
	}
	if err := uow.LoreEntries().Save(ctx, updated); err != nil {
		return domain.LoreEntry{}, err
	}
	if audit != nil {
		if err := uow.Audits().Add(ctx, *audit); err != nil {
			return domain.LoreEntry{}, err
		}
	}
	if err := uow.Commit(ctx); err != nil {
		return domain.LoreEntry{}, err
	}
	return updated, nil
}

func isNotFound(err error, target **domain.NotFoundError) bool {
	if err == nil {
		return false
	}
	if nf, ok := err.(*domain.NotFoundError); ok {
		*target = nf
		return true
	}
	return false
}

File: src/memlore/application/commands/verify_lore.py
from __future__ import annotations

from dataclasses import dataclass

from memlore.application.ports.clock import Clock
from memlore.application.ports.unit_of_work import UnitOfWork
from memlore.domain.exceptions import NotFoundError, ValidationError
from memlore.domain.models.lore_entry import LoreEntry
from memlore.domain.services.verification import apply_verification


@dataclass(frozen=True, slots=True)
class VerifyLoreCommand:
    entry_id: str
    actor_id: str


class VerifyLoreHandler:
    def __init__(self, uow: UnitOfWork, clock: Clock) -> None:
        self._uow = uow
        self._clock = clock

    def handle(self, command: VerifyLoreCommand) -> LoreEntry:
        actor = command.actor_id.strip()
        if not actor:
            raise ValidationError("actor must be non-empty")

        entry = self._uow.lore_entries.get(command.entry_id)
        if entry is None:
            raise NotFoundError(f"lore entry {command.entry_id} not found")

        updated, audit = apply_verification(
            entry, actor_id=actor, now=self._clock.now()
        )
        self._uow.lore_entries.save(updated)
        if audit is not None:
            self._uow.audits.add(audit)
        self._uow.commit()
        return updated
